components/script/dom/readablestreamdefaultreader.rs

```diff
@@ -0,0 +1,488 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::cell::Cell;
+use std::collections::VecDeque;
+use std::mem;
+use std::rc::Rc;
+
+use dom_struct::dom_struct;
+use js::jsapi::Heap;
+use js::jsval::{JSVal, UndefinedValue};
+use js::rust::{HandleObject as SafeHandleObject, HandleValue as SafeHandleValue};
+
+use super::bindings::refcounted::Trusted;
+use super::bindings::root::MutNullableDom;
+use super::types::ReadableStreamDefaultController;
+use crate::dom::bindings::cell::DomRefCell;
+use crate::dom::bindings::codegen::Bindings::ReadableStreamDefaultReaderBinding::{
+    ReadableStreamDefaultReaderMethods, ReadableStreamReadResult,
+};
+use crate::dom::bindings::error::Error;
+use crate::dom::bindings::import::module::Fallible;
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject, Reflector};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::trace::RootedTraceableBox;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::promise::Promise;
+use crate::dom::promisenativehandler::{Callback, PromiseNativeHandler};
+use crate::dom::readablestream::ReadableStream;
+use crate::dom::teereadrequest::TeeReadRequest;
+use crate::realms::{enter_realm, InRealm};
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+/// <https://streams.spec.whatwg.org/#read-request>
+/// For now only one variant: the one matching a `read` call.
+#[derive(JSTraceable)]
+pub enum ReadRequest {
+    /// <https://streams.spec.whatwg.org/#default-reader-read>
+    Read(Rc<Promise>),
+    /// <https://streams.spec.whatwg.org/#ref-for-read-request%E2%91%A2>
+    Tee {
```

Comment:
`DefaultTee` would be more precise, but that can be done when bytes stream is implemented.